block/retriever.go

```diff
@@ -1,38 +1,33 @@
 package block
 
 import (
-	"context"
 	"fmt"
 
-	"code.cloudfoundry.org/go-diodes"
-
 	"github.com/dymensionxyz/dymint/da"
+	"github.com/dymensionxyz/dymint/settlement"
+	"github.com/tendermint/tendermint/libs/pubsub"
 )
 
-// RetrieveLoop listens for new target sync heights and then syncs the chain by
-// fetching batches from the settlement layer and then fetching the actual blocks
-// from the DA.
-func (m *Manager) RetrieveLoop(ctx context.Context) (err error) {
-	m.logger.Info("Started retrieve loop.")
-	p := diodes.NewPoller(m.targetSyncHeight, diodes.WithPollingContext(ctx))
-
-	for {
-		targetHeight := p.Next() // We only care about the latest one
-		if targetHeight == nil {
-			return
-		}
-
-		if err = m.syncToTargetHeight(*(*uint64)(targetHeight)); err != nil {
-			err = fmt.Errorf("sync until target: %w", err)
-			return
-		}
+// onReceivedNewBatch will take a block and apply it
+func (m *Manager) onReceivedBatch(event pubsub.Message) {
+	eventData, ok := event.Data().(*settlement.EventDataNewBatchAccepted)
+	if !ok {
+		m.logger.Error("onReceivedBatch", "err", "wrong event data received")
+		return
+	}
+	h := eventData.EndHeight
+	err := m.syncToTargetHeight(h)
+	if err != nil {
+		m.logger.Error("sync until target", "err", err)
 	}
 }
 
 // syncToTargetHeight syncs blocks until the target height is reached.
 // It fetches the batches from the settlement, gets the DA height and gets
 // the actual blocks from the DA.
 func (m *Manager) syncToTargetHeight(targetHeight uint64) error {
+	defer m.syncFromDaMu.Unlock()
+	m.syncFromDaMu.Lock()
```

Comment:
I dont get why this is here and not in syncFromDABatch

Reply:
its to prevent concurrent syncing processes when a there is a new batch notification, but still not finished from previous syncing. i think its better here.

Reply:
Commented also above that I don't think it does anything (as seems like no 2 concurrent calls to this function are being executed).

Reply:
there might be two calls for it, but both `syncFromDABatch` and `applyLocalBlock` are already have mutex inside, so not sure new `syncFromDaMu` is needed